# Q1


*   In this question we want to use POS-tagged training set to compute for each word the tag that maximizes $p(t|w)$.
*   We will implement a simple tokenizer to deal with sentence boundaries.
*   We start by assuming that all unknown words are NN and compute error rate on known and unknown words.
*   Then write at least five rules to do a better job of tagging unknown words, and show the difference in error rates.


In [1]:
import re
import math

import nltk
nltk.download('brown')
from nltk.corpus import brown

[nltk_data] Downloading package brown to /root/nltk_data...
[nltk_data]   Unzipping corpora/brown.zip.


In [2]:
def generate_dict(_samples):
    """
    Generate a dictionary that captures the count of each (word, tag) combination from a list of samples.

    Args:
    _samples (list): List of tuples containing (word, tag) pairs.

    Returns:
    dict: A dictionary where keys are words and values are lists of dictionaries with 'tag' and 'count' keys.
          Each dictionary in the list represents a unique tag associated with the word and its count.
    """

    '''
    Example:
      _samples = [('apple', 'fruit'), ('banana', 'fruit'), ('apple', 'fruit'), ('apple', 'color'), ('banana', 'color')]
      After calling generate_dict(_samples), the expected output would be:
      {
          'apple': [
              {'tag': 'fruit', 'count': 2},
              {'tag': 'color', 'count': 1}
          ],
          'banana': [
              {'tag': 'fruit', 'count': 1},
              {'tag': 'color', 'count': 1}
          ]
      }
      This represents that 'apple' appeared twice with the tag 'fruit' and once with the tag 'color',
      while 'banana' appeared once with both 'fruit' and 'color' tags.

    '''

    dictionary = {}


    ## Your code here
    # create the dictionary described in the comments
    for word, tag in _samples:
        if word not in dictionary:
            dictionary[word] = []

        # Check if the tag already exists for the word
        tag_exists = False
        for entry in dictionary[word]:
            if entry['tag'] == tag:
                entry['count'] += 1
                tag_exists = True
                break

        # If the tag doesn't exist, add a new entry with count 1
        if not tag_exists:
            dictionary[word].append({'tag': tag, 'count': 1})
    ## End Code
    def get_tag_counts(subitem):
        return subitem['count']

    for item in dictionary:
        sorted(dictionary[item], key=get_tag_counts, reverse=True)

    return dictionary

In [3]:
def predict_tag(_test_set, _tag_dict):
    """
    Predicts the tags for a given test set of words based on a provided tag dictionary.

    Args:
    _test_set (list): A list of tuples containing (word, true_tag) pairs to be predicted.
    _tag_dict (dict): A dictionary containing words as keys and lists of dictionaries with 'tag' and 'count' keys as values.

    Returns:
    float: The accuracy of the predictions, calculated as the ratio of correct predictions to the total number of predictions.

    Comments:
    - For unknown words, 'NN' (noun) tag is assigned.
    - Tags are assigned based on the highest count for known words.
      - If there are more than 1 tag for a given word, the tag with the highest count is chosen.
      - If there is only 1 tag available, it is returned directly.
    """

    accuracy = 0
    for item in _test_set:
        word = item[0]
        true_tag = item[1]
        if word in _tag_dict:
            # Predict the tag based on the highest count
            prediction = _tag_dict[word][0]['tag'] ## Your code here
        else:
            # Assign 'NN' tag for unknown words
            prediction = 'NN' ## Your code here
        if prediction == true_tag:
          accuracy = accuracy + 1 ## Your code here

    accuracy /= len(_test_set)
    print(f"Assuming that all unknown words are NN")
    print(f">> accuracy: {accuracy}")
    return accuracy


In [4]:
def predict_tag_with_improvements(_test_set, _tag_dict):
    """
    Predicts the tags for a given test set of words based on a provided tag dictionary, with additional rules for unknown words.

    Args:
    _test_set (list): A list of tuples containing (word, true_tag) pairs to be predicted.
    _tag_dict (dict): A dictionary containing words as keys and lists of dictionaries with 'tag' and 'count' keys as values.

    Returns:
    float: The accuracy of the predictions, calculated as the ratio of correct predictions to the total number of predictions.

    Comments:
    - For unknown words, 'NN' (noun) tag is initially assigned.
    - Additional rules are applied to analyze unknown words and assign more specific tags based on patterns observed in the word:
        - 'VBG' (verb, gerund) for words ending in 'ing'
        - 'NP$' (noun, possessive) for words ending in "'s"
        - 'NNS' (noun, plural) for words ending in 's'
        - 'RB' (adverb) for words ending in 'ly'
        - 'VBN' (verb, past participle) for words ending in 'ed'
        - 'JJ' (adjective) for words matching certain patterns like 'ble', 'ish', 'ful', etc.
        - 'CD' (cardinal numeral) for numeric strings
        - 'NP' (noun, proper singular) for capitalized words
    """

    ## Your code here
    accuracy = 0
    for item in _test_set:
        word = item[0]
        true_tag = item[1]

        if word in _tag_dict:
            # Predict the tag based on the highest count
            prediction = _tag_dict[word][0]['tag']
        else:
            # Assign 'NN' tag for unknown words
            prediction = 'NN'

            # Apply additional rules for unknown words
            if word.endswith('ing'):
                prediction = 'VBG'
            elif word.endswith("'s"):
                prediction = 'NP$'
            elif word.endswith('s'):
                prediction = 'NNS'
            elif word.endswith('ly'):
                prediction = 'RB'
            elif word.endswith('ed'):
                prediction = 'VBN'
            elif any(pattern in word for pattern in ['ble', 'ish', 'ful']):
                prediction = 'JJ'
            elif word.isdigit():
                prediction = 'CD'
            elif word[0].isupper():
                prediction = 'NP'

        if prediction == true_tag:
            accuracy += 1
    ## End Code
    accuracy /= len(_test_set)
    print(f"With additional rules for unknown words")
    print(f">> accuracy: {accuracy}")
    return accuracy

In [7]:
CORPUS = brown.tagged_words(categories='news')
CORPUS_SIZE = len(brown.tagged_words(categories='news'))

CUT_OFF = math.floor(CORPUS_SIZE * 0.75)

# section off training and testing lists from corpus
training_list = CORPUS[:CUT_OFF]
testing_list = CORPUS[CUT_OFF:]

# duplicates are ignored in sets
training_set = set(training_list)
testing_set = set(testing_list)
intersection = training_set.intersection(testing_set)

print(f"length of training set:     {len(training_list)}")
print(f"length of testing set:      {len(testing_list)}")

# uncomment to see how much the training set and testing set overlap
print(f"intersection:               {len(intersection)}")

# uncomment to survey tagged corpus
# print(training_set)

tag_dict = generate_dict(training_list)
accurary_base = predict_tag(testing_list, tag_dict)
accurary_impr = predict_tag_with_improvements(testing_list, tag_dict)
delta = math.floor((accurary_impr - accurary_base) * len(testing_list))
print(f"{delta} more words got correctly classified.")


length of training set:     75415
length of testing set:      25139
intersection:               3429
Assuming that all unknown words are NN
>> accuracy: 0.8184096423883209
With additional rules for unknown words
>> accuracy: 0.8626039221926091
1110 more words got correctly classified.


In this code snippet, a corpus of tagged words from the 'news' category of the Brown corpus is used to train and test a tag prediction model. The corpus is split into a training set and a testing set based on a specified cutoff point.

The training set is used to generate a tag dictionary using the `generate_dict` function. This dictionary captures the count of each (word, tag) combination in the training set.

The testing set is then used to evaluate the accuracy of the tag prediction model. The `predict_tag` function is called to predict the tags for the words in the testing set based on the tag dictionary generated from the training set. The accuracy of the predictions is calculated as the ratio of correct predictions to the total number of predictions.

The process is repeated with the `predict_tag_with_improvements` function, which applies additional rules for predicting tags for unknown words. The accuracy of the improved predictions is also calculated.

The results of the code execution are as follows:
- Length of the training set: 75,415 words
- Length of the testing set: 25,139 words
- Intersection between the training and testing sets: 3,429 words (words that appear in both sets)

The accuracy of the tag predictions without the additional rules for unknown words is approximately 0.8184 (81.84%). This means that around 81.84% of the predicted tags match the true tags in the testing set.

After applying the additional rules for unknown words, the accuracy of the predictions improves to approximately 0.8626 (86.26%). This means that around 86.26% of the predicted tags match the true tags in the testing set. The improvement in accuracy is significant, with 1,110 more words correctly classified compared to the base model without the additional rules.

The analysis suggests that the additional rules for unknown words help to assign more specific tags based on observed patterns in the words. This leads to a higher accuracy in predicting the correct tags for unknown words in the testing set.